File: tests/src/commonTest/kotlin/primitives/CallbackReturnPrimitive.kt
package primitives

import natives.test.MyDictionary
import natives.test.MyEnum
import natives.test.VoidCallback
import withLib
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class CallbackReturnPrimitive {

    @Test
    fun callbackReturnChar() = withLib {
        assertTrue(natives.test.callbackReturnChar { 'a' })
    }

    @Test
    fun callbackReturnBoolean() = withLib {
        assertTrue(natives.test.callbackReturnBoolean { true })
    }

    @Test
    fun callbackReturnByte() = withLib {
        assertTrue(natives.test.callbackReturnByte { 1.toByte() })
    }

    @Test
    fun callbackReturnShort() = withLib {
        assertTrue(natives.test.callbackReturnShort { 1.toShort() })
    }

    @Test
    fun callbackReturnInt() = withLib {
        assertTrue(natives.test.callbackReturnInt { 1 })
    }

    @Test
    fun callbackReturnLong() = withLib {
        assertTrue(natives.test.callbackReturnLong { 1.toLong() })
    }

    @Test
    fun callbackReturnFloat() = withLib {
        assertTrue(natives.test.callbackReturnFloat { 1.1f })
    }

    @Test
    fun callbackReturnDouble() = withLib {
        assertTrue(natives.test.callbackReturnDouble { 1.1 })
    }

    @Test
    fun callbackReturnString() = withLib {
        assertTrue(natives.test.callbackReturnString { "test string" })
    }

    @Test
    fun callbackReturnCallback() = withLib {
        val toPass = VoidCallback {}
        assertEquals(toPass, natives.test.callbackReturnCallback { toPass })
    }

    @Test
    fun callbackReturnEnum() = withLib {
        assertTrue(natives.test.callbackReturnEnum { MyEnum.CASE2 })
    }

    @Test
    fun callbackReturnDictionary() = withLib {
        assertTrue(natives.test.callbackReturnDictionary { MyDictionary(1, 2, 3, 4) })
    }
}

File: tests/natives/test/src/primitives/callback_return_primitive.c
#include <api.h>
#include <string.h>

KBoolean callbackReturnChar(CallbackReturnChar* arg) {
    return arg->invoke(arg) == 'a';
}

KBoolean callbackReturnBoolean(CallbackReturnBoolean* arg) {
    return arg->invoke(arg) == true;
}

KBoolean callbackReturnByte(CallbackReturnByte* arg) {
    return arg->invoke(arg) == 1;
}

KBoolean callbackReturnShort(CallbackReturnShort* arg) {
    return arg->invoke(arg) == 1;
}

KBoolean callbackReturnInt(CallbackReturnInt* arg) {
    return arg->invoke(arg) == 1;
}

KBoolean callbackReturnLong(CallbackReturnLong* arg) {
    return arg->invoke(arg) == 1;
}

KBoolean callbackReturnFloat(CallbackReturnFloat* arg) {
    return arg->invoke(arg) == 1.1f;
}

KBoolean callbackReturnDouble(CallbackReturnDouble* arg) {
    return arg->invoke(arg) == 1.1;
}

KBoolean callbackReturnString(CallbackReturnString* arg) {
    return strcmp(arg->invoke(arg).data, "test string") == 0;
}

VoidCallback* callbackReturnCallback(CallbackReturnCallback* arg) {
    return arg->invoke(arg);
}

KBoolean callbackReturnEnum(CallbackReturnEnum* arg) {
    return arg->invoke(arg) == MyEnum_CASE2;
}

KBoolean callbackReturnStruct(CallbackReturnStruct* arg) {
    MyStruct result = arg->invoke(arg);
    return result.a == 1 &&
        result.b == 2 &&
        result.c == 3 &&
        result.d == 4;
}